### PR TITLE
Documentation Parser Extended Integration

### PR DESCRIPTION
**Pull request by BinarySoftware**
## TO DO LIST
 - [x] Parsing code containing comment with function defined in next line -
   nothing more complicated. In result we want documentation just like we get
    now, but with title created from function name. After updating code to
    newest parser ( lets assume wdanilo/cata branch is your reference now on)
    make a PR onto it. It may be dirty here and there, but it should work

-  [x] Add to our AST node "documented" which will contain two fields - other AST
    and your Documentation. Your doc shouldn't extend AST ( it doesn't on my
    branch already ) - separate PR 
   ( RE: Cannot be in separate PR as it is vital for functionality of Doc Parser creator )

-  [x] Now use case working like that - in code we have `def Maybe` with doc
    above it and to every function in it (refer to parser tests). We want to
    traverse it and make final form of documentation. Take from def name of
    documentation, show documentation of every internal function just like
    apple displays on grey background in Developer.Apple.com Documentation.
    Generally implementation here should be extremely easy as def is block
    which contains list of lines so you loop through it and if you find Doc
    then check if next line is empty or contain function ( and you have it
    already from point 1 )

#### For reference - (#121)
This is really important, as it deeply integrates together DocParser with Parser. Now if you have some code like that:
```
## This function adds together x and y
foo x y = x + y
```
it will output to you:
```
Documentation(Documented(This function adds together x and y),AST(...))
```
In short words - it combines together parser output with docParser output in a better way, creates titles for documentation out of function names.

### Important Notes
- As it is a small PR it can only create title from function name, some methods may be generalised in future

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 1]
```diff
@@ -1,51 +1,320 @@
...
+import flexer.Parser.{Result => res}
+import org.enso.data.List1
+import org.enso.syntax.text.AST.Block.{LineOf => Line}
 
 ////////////////////////////////////////////////////////////////////////////////
-//// DocParser /////////////////////////////////////////////////////////////////
+//// Doc Parser ////////////////////////////////////////////////////////////////
 ////////////////////////////////////////////////////////////////////////////////
 
+/** This is the main Parser class. It is being used to create documentation from
```
Comment (iamrecursion) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: This comment is worded a little strangely. Perhaps something like "it is used to create structured documentation from the blocks of comment text created by the main parser" or similar.

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 2]
```diff
@@ -1,51 +1,320 @@
...
+  * This is Doc Parser Runner. Essentially it binds together Enso Parser with
+  * Doc Parser. When Parser finishes its job it invokes runner with its created
+  * AST with resolved macros. Then Doc Parser Runner does it's job - running Doc
+  * Parser on every [[AST.Comment]], combined with connecting Doc with AST in
+  * Documentation node, getting AST from Def's and Infix'es
+  */
+object DocParserRunner {
+
+  /**
+    * function for invoking DocParser in right places
```
Comment (iamrecursion) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: This comment is really strangely formatted. It should be written as a paragraph not a series of sentences. 

```
  /**
    * This function invokes the documentation parser on every `Comment`.
    * 
    * It generates documentation for every comment before traversing the AST in
    * order to insert these. It then uses this information to generate HTML
    * files containing the human-readable documentation before returning the
    * modified AST.
    * ...
```

Please don't just copy and paste this. I don't know the details of ScalaDoc's formatting so you may need to alter it to work properly.

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 3]
```diff
@@ -1,51 +1,320 @@
...
+    * After all that it generates HTML files for created Documented's
+    * and outputs modified AST
+    *
+    * @param module - parsed data by Parser
+    * @return - AST with possible documentation
+    */
+  def document(module: AST.Module): AST = {
+    val createdDocs = createDocs(module)
+
+    /** NOTE : Comment out for ease of debugging procedures
```
Comment (iamrecursion) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: This should be a source note if it must stay, but this should not be in a PR against master. If this is a problem, then the HTML generation behaviour needs to be configurable at runtime.

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 4]
```diff
@@ -1,51 +1,320 @@
...
+      */
+    //generateHTMLForEveryDocumented(createdDocs)
+    createdDocs
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  //// Created Doc's in right places with appropriate AST //////////////////////
+  //////////////////////////////////////////////////////////////////////////////
+  /**
+    * This function gets [[AST.Module]] or [[AST.Block]]
```
Comment (iamrecursion) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Why is this comment not using the entire width of the code? Furthermore, it's not well-written.

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 5]
```diff
@@ -1,51 +1,320 @@
...
+  }
+
+  /**
+    * this is a helper function for creating docs with AST.
+    * Essentially it traverses through lines and tries to find a pattern on them
+    *
+    * @param lines - AST lines
+    * @return - lines with possibly Doc with added AST
+    */
+  //FIXME - Add offset to AST line
```
Comment (iamrecursion) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Please follow the conventions for `FIXME` comments as set out in the style guide.

File: Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala [hunk 6]
```diff
@@ -17,7 +17,7 @@ import org.scalatest.Assertion
 //    output match {
```
Comment (iamrecursion) [inline Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: Tests that are commented out should not make it onto master. Either the tests need to be fixed, or they need to be made `pending` with accompanying tasks to make them work.

Comment (BinarySoftware) [reply Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: These tests are working, but Wojtek have commented them out for some purpose, and he doesn't want me to uncomment those because then the PR would become too big

Comment (iamrecursion) [reply Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: If the tests are functional then they should be uncommented on master.

Comment (BinarySoftware) [reply Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: Then I don't know what to do as Wojtek has told me not to uncomment those tests and to only do what I've been requested to do. @wdanilo what should I do? Can I uncomment them??

Comment (iamrecursion) [reply Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: I'm particularly interested as to why _working_ tests are commented out. No PR should be accepted without functional tests. It's part of our acceptance criteria.

Comment (BinarySoftware) [reply Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: Then why parser has been merged into master with my tests commented out

Comment (iamrecursion) [reply Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: I have no idea. The parser merge is a special case and many rules have been relaxed in the name of expediency, but if the tests were _functional_ then I have no idea.

Comment (BinarySoftware) [reply Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: And they are totally functional, ~60 working and passing tests

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 7]
```diff
@@ -1,51 +1,336 @@
...
+    * human-readable documentation before returning the modified AST to Parser.
+    *
+    * @param module - parsed data by Parser
+    * @return - AST with possible documentation
+    */
+  def document(module: AST.Module, withHTMLGeneration: Boolean): AST = {
+    val createdDocs = createDocs(module)
+    if (withHTMLGeneration) generateHTMLForEveryDocumented(createdDocs)
+    createdDocs
+  }
```
Comment (wdanilo) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: This function always discards the line `if (withHTMLGeneration) generateHTMLForEveryDocumented(createdDocs)`, I think its not expected behavior.

Plus, I would not try to create a single API entry point in such a case . If you've got a function that adds fdocumentation and other that traverses AST nad outputs HTML from it, lets make two separate public functions in the API for that.

Plus, a public generateHTML function should not save the HTML to file - it should return the HTML as an output. Then we can save it out to the prefered destination. Its much more flexible than now.

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 8]
```diff
@@ -1,51 +1,336 @@
...
+    * through their lines to create [[AST.Documented]] from [[AST.Comment]] if
+    * found, with or without AST from [[AST.App.Infix]] or [[AST.Def]]
+    *
+    * @param ast - module with possibility to create Documentation from comments
+    * @return - modified data containing possibly Documentation(s) with AST
+    */
+  def createDocs(ast: AST): AST = {
+    ast match {
+      case AST.Module.any(m) =>
+        traverseThroughModule(m)
```
Comment (wdanilo) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: This function does more things than just travese, right? Wouldn't it be better to just name it `createDocsForModule` or something along these lines?

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 9]
```diff
@@ -1,51 +1,336 @@
...
+   *  should start from the same place as it has started before.
+   */
+  /**
+    * this is a helper function for creating docs with AST.
+    * Essentially it traverses through lines and tries to find a pattern on them
+    *
+    * @param lines - AST lines
+    * @return - lines with possibly Doc with added AST
+    */
+  def transformLines(lines: List[AST.Block.OptLine]): List[AST.Block.OptLine] =
```
Comment (wdanilo) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: I'd find more descriptine name here, like, attachDocToSubsequentAST? Or anythinfg else, but `transformLines` does not reall tell what it does.

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 10]
```diff
@@ -1,51 +1,336 @@
...
+    lines match {
+      case ::(line1, tail) =>
```
Comment (wdanilo) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Why we use this instead of just `line1 :: tail` ?

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 11]
```diff
@@ -1,51 +1,336 @@
...
+        line1 match {
+          case Line(Some(AST.Comment.any(com)), comOff) =>
+            tail match {
+              case ::(line2, rest) =>
+                line2 match {
+                  case Line(Some(AST.App.Infix.any(ast)), astOff) =>
+                    createDocumentedLine(com, Some(ast), comOff, astOff) :: transformLines(
+                      rest
+                    )
```
Comment (wdanilo) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: partition every multiline expr to separate variables :)

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 12]
```diff
@@ -1,51 +1,336 @@
...
+                  case Line(Some(AST.Def.any(ast)), astOff) =>
+                    createDocumentedLine(
+                      com,
+                      Some(createDocs(ast)),
+                      comOff,
+                      astOff
+                    ) :: transformLines(rest)
```
Comment (wdanilo) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: partition every multiline expr to separate variables :)

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/Parser.scala [hunk 13]
```diff
@@ -293,10 +293,27 @@ object Main extends App {
...
+  val in3  = "(a) b = c"
+  val in4  = "if a then (b)"
+  val in2  = "(a) b = c]"
+  val inp2 = "a (b (c)) x"
+
+  val inp = """## This function adds *x* to *y*
+              |add x y = x + y
+              |mul x y = x * y
+              |
+              |## This function divides *x* by *y*
```
Comment (wdanilo) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/Parser.scala]: It should be 
```
## This function divides `x` by `y`
```

File: Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala (hunk 6 above)
Comment (wdanilo) [reply Syntax/specialization/src/test/scala/org/enso/syntax/text/DocParserTests.scala]: I just asked to keep PR relative to changes. So it we want to do changes that are not relative to this PR, lets do them in a separate PR. Ara is right though that we should not have things commented out, so if these tests are working properly, lets just create a seaparate PR enabling them and merge it fast!

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 14]
```diff
@@ -1,51 +1,294 @@
...
+import flexer.Parser.{Result => res}
+import org.enso.data.List1
+import org.enso.syntax.text.AST.Block.{LineOf => Line}
 
 ////////////////////////////////////////////////////////////////////////////////
-//// DocParser /////////////////////////////////////////////////////////////////
+//// Doc Parser ////////////////////////////////////////////////////////////////
 ////////////////////////////////////////////////////////////////////////////////
 
+/** This is the class used to invoke Documentation Parser.
```
Comment (iamrecursion) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]: Please leave a blank line between the summary line and the main paragraph.

File: Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 15]
```diff
@@ -1,51 +1,294 @@
...
+  * it is used to create structured documentation from the blocks of commented
```
Comment (iamrecursion) [inline Syntax/specialization/src/main/scala/org/enso/syntax/text/DocParser.scala]:
```suggestion
  * It is used to create structured documentation from the blocks of commented
```
